How many datasets have CC0 (or CCBY for forked repositories) and all 8 "Terms of Use" fields filled? How many have 7? How many have 6? Etc...

In [153]:
import csv
import numpy as np
import pandas as pd


In [154]:
data = pd.read_csv('terms_metadata.tab', sep='\t', na_filter = False)

In [155]:
# Get only metadata for the latest versions of each dataset
latestversion = data.iloc[data.groupby('persistentUrl')['datasetVersionId'].agg(pd.Series.idxmax)].sort_values(by=['publisher'], inplace=False, ascending=True).reset_index(drop=True, inplace=False)


In [156]:
# Replace any blank values with NaN
latestversion = latestversion.replace(r'^\s*$', np.nan, regex=True)

# Replace publsiher "Root" with Dartmouth
latestversion['publisher'] = latestversion['publisher'].replace(['Root'],'Dartmouth')


In [157]:
latestversion_termsofuse = latestversion.drop(columns=[
    'publisher', 'datasetVersionId', 'majorVersionNumber', 'minorVersionNumber', 'termsOfUse',
    'termsOfAccess', 'availabilityStatus', 'contactForAccess', 'dataaccessPlace',
    'originalArchive', 'sizeOfCollection', 'studyCompletion'])


In [158]:
latestversion_termsofuse.license.unique()

array(['NONE', 'CC0', 'CC BY', nan, 'CCBY'], dtype=object)

In [159]:
latestversionCC0 = (latestversion_termsofuse
    .query('license == "CC0" or license == "CCBY" or license == "CC BY"')
#     .reset_index(drop = False, inplace = True)
    .set_index('persistentUrl')
)

In [160]:
latestversionCC0.head(5)

license citationRequirements conditions  \
persistentUrl                                                                   
http://dx.doi.org/10.4225/87/B5AJXD       CC0                  NaN        NaN   
https://hdl.handle.net/11529/10548094     CC0                  NaN        NaN   
https://hdl.handle.net/11529/10548095     CC0                  NaN        NaN   
https://hdl.handle.net/11529/10548096     CC0                  NaN        NaN   
https://hdl.handle.net/11529/10548097     CC0                  NaN        NaN   

                                      confidentialityDeclaration  \
persistentUrl                                                      
http://dx.doi.org/10.4225/87/B5AJXD                          NaN   
https://hdl.handle.net/11529/10548094                        NaN   
https://hdl.handle.net/11529/10548095                        NaN   
https://hdl.handle.net/11529/10548096                        NaN   
https://hdl.handle.net/11529/10548097                        NaN   

                                      depositorRequirements disclaimer  \
persistentUrl                                                            
http://dx.doi.org/10.4225/87/B5AJXD                     NaN        NaN   
https://hdl.handle.net/11529/10548094                   NaN        NaN   
https://hdl.handle.net/11529/10548095                   NaN        NaN   
https://hdl.handle.net/11529/10548096                   NaN        NaN   
https://hdl.handle.net/11529/10548097                   NaN        NaN   

                                      restrictions specialPermissions  
persistentUrl                                                          
http://dx.doi.org/10.4225/87/B5AJXD            NaN                NaN  
https://hdl.handle.net/11529/10548094          NaN                NaN  
https://hdl.handle.net/11529/10548095          NaN                NaN  
https://hdl.handle.net/11529/10548096          NaN                NaN  
https://hdl.handle.net/11529/10548097          NaN                NaN

In [161]:
len(latestversionCC0)

33961

In [162]:
countsByDataset = (latestversionCC0
    .drop(columns=['license'])
    .count(axis=1)
    .to_frame('count')
    .query('count > 0')
)

In [163]:
countsByDataset.head(5)

count
persistentUrl                            
https://doi.org/10.21979/N9/OF5ZDK      2
https://doi.org/10.21979/N9/DHYM9H      2
https://doi.org/10.34894/FUTGYT         3
https://doi.org/10.34894/Q0XNJS         1
https://doi.org/10.34894/AXFRPB         2

In [164]:
len(countsByDataset)

498

Looks like there are 498 datasets with:
- Either the CC0 waiver or CCBY license
- One or more Terms of Use fields filled (exluding the Terms of Use field)

Let's join the countsByDataset dataframe to the latestversionCC0 dataframe: 

In [176]:
countsByDataset_merged = (pd
    .merge(countsByDataset, latestversion, how='left', on='persistentUrl')
    .drop(columns=[
        'termsOfUse', 'termsOfAccess', 'availabilityStatus',
        'contactForAccess', 'dataaccessPlace', 'originalArchive', 'sizeOfCollection', 'studyCompletion'])
)


In [177]:
countsByDataset_merged.head(5)

persistentUrl  count      publisher  datasetVersionId  \
0  https://doi.org/10.21979/N9/OF5ZDK      2  DR-NTU (Data)              1432   
1  https://doi.org/10.21979/N9/DHYM9H      2  DR-NTU (Data)              1431   
2     https://doi.org/10.34894/FUTGYT      3    DataverseNL              2997   
3     https://doi.org/10.34894/Q0XNJS      1    DataverseNL              1412   
4     https://doi.org/10.34894/AXFRPB      2    DataverseNL              1877   

   majorVersionNumber  minorVersionNumber license  \
0                   1                   0     CC0   
1                   1                   0     CC0   
2                   2                   0     CC0   
3                   1                   0     CC0   
4                   1                   2     CC0   

                                citationRequirements  \
0  To cite this dataset: \nStyles, Suzy; Travers ...   
1  To cite this dataset:\nStyles, Suzy J; Bin Mus...   
2  Mention publication Habets et al 2020 when usi...   
3                                                NaN   
4  If active involvement of AIGHD/IFPRI researche...   

                                          conditions  \
0                                                NaN   
1                                                NaN   
2  Anonimyzed by changing patients' age in age ra...   
3  The data, syntaxes, and reports of this projec...   
4                                                NaN   

  confidentialityDeclaration depositorRequirements disclaimer  \
0                        NaN                   NaN        NaN   
1                        NaN                   NaN        NaN   
2                        NaN                   NaN        NaN   
3                        NaN                   NaN        NaN   
4                        NaN                   NaN        NaN   

                                        restrictions  \
0  Researchers must give their name and current r...   
1  To download these audio files, researchers mus...   
2                  Not available for commercial use.   
3                                                NaN   
4                                                NaN   

                                  specialPermissions  
0                                                NaN  
1                                                NaN  
2                                                NaN  
3                                                NaN  
4  The data are available upon request through th...

In [178]:
len(countsByDataset_merged)

498

In [179]:
# Export to CSV
file = 'countsByDataset_merged.csv'
countsByDataset_merged.to_csv(file, index=False)